File: Raven.Client.Java/lib/src/main/java/raven/client/connection/HttpJsonRequestFactory.java
package raven.client.connection;

import java.util.ArrayList;
import java.util.Date;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.atomic.AtomicInteger;

import org.apache.commons.lang.StringUtils;
import org.apache.http.HttpStatus;
import org.apache.http.client.HttpClient;
import org.apache.http.impl.client.DecompressingHttpClient;
import org.apache.http.impl.client.DefaultHttpClient;
import org.apache.http.impl.client.StandardHttpRequestRetryHandler;

import raven.abstractions.basic.EventHandler;
import raven.abstractions.basic.EventHelper;
import raven.abstractions.connection.WebRequestEventArgs;
import raven.abstractions.connection.profiling.RequestResultArgs;
import raven.abstractions.data.Constants;
import raven.abstractions.data.HttpMethods;
import raven.abstractions.json.linq.RavenJToken;
import raven.client.extensions.MultiDatabase;
import raven.client.util.SimpleCache;

/**
 * Create the HTTP Json Requests to the RavenDB Server
 * and manages the http cache
 */
public class HttpJsonRequestFactory implements AutoCloseable {

  private HttpClient httpClient;

  private List<EventHandler<WebRequestEventArgs>> configureRequest = new ArrayList<>();

  private List<EventHandler<RequestResultArgs>> logRequest = new ArrayList<>();

  private final int maxNumberOfCachedRequests;
  private SimpleCache cache;
  protected AtomicInteger numOfCachedRequests = new AtomicInteger();
  protected int numOfCacheResets;
  private boolean disableRequestCompression;
  private boolean enableBasicAuthenticationOverUnsecuredHttpEvenThoughPasswordsWouldBeSentOverTheWireInClearTextToBeStolenByHackers;
  private ThreadLocal<Long> aggressiveCacheDuration = new ThreadLocal<>(); // in milis
  private ThreadLocal<Boolean> disableHttpCaching = new ThreadLocal<>();
  private volatile boolean disposed;

  public HttpJsonRequestFactory(int maxNumberOfCachedRequests) {
    super();
    DefaultHttpClient innerClient = new DefaultHttpClient();
    this.httpClient = new DecompressingHttpClient(innerClient);
    innerClient.setHttpRequestRetryHandler(new StandardHttpRequestRetryHandler(0, false));
    this.maxNumberOfCachedRequests = maxNumberOfCachedRequests;
    resetCache();
  }


  public void addConfgureRequestEventHandler(EventHandler<WebRequestEventArgs> event) {
    configureRequest.add(event);
  }


  public void addLogRequestEventHandler(EventHandler<RequestResultArgs> event) {
    logRequest.add(event);
  }

  /**
   * @return the httpClient
   */
  public HttpClient getHttpClient() {
    return httpClient;
  }

  protected void cacheResponse(String url, RavenJToken data, Map<String, String> headers) {
    if (StringUtils.isEmpty(headers.get("ETag"))) {
      return;
    }

    RavenJToken clone = data.cloneToken();
    clone.ensureCannotBeChangeAndEnableShapshotting();

    cache.set(url, new CachedRequest(clone, new Date(), new HashMap<>(headers), MultiDatabase.getDatabaseName(url), false));

  }

  @Override
  public void close() throws Exception {
    if (disposed) {
      return ;
    }
    disposed = true;
    cache.close();
  }

  private CachedRequestOp configureCaching(String url, HttpJsonRequest request) {
    CachedRequest cachedRequest = cache.get(url);
    if (cachedRequest == null) {
      return new CachedRequestOp(null, false);
    }
    boolean skipServerCheck = false;
    if (getAggressiveCacheDuration() != null) {
      long totalSeconds = getAggressiveCacheDuration() / 1000;
      if (totalSeconds > 0) {
        request.getWebRequest().addHeader("Cache-Control", "max-age=" + totalSeconds);
      }

      if (cachedRequest.isForceServerCheck() == false && (new Date().getTime() - cachedRequest.getTime().getTime() < getAggressiveCacheDuration())) { //can serve directly from local cache
        skipServerCheck = true;
      }
      cachedRequest.setForceServerCheck(false);
    }
    request.getWebRequest().addHeader("If-None-Match", cachedRequest.getHeaders().get("ETag"));
    return new CachedRequestOp(cachedRequest, skipServerCheck);
  }

  public HttpJsonRequest createHttpJsonRequest(CreateHttpJsonRequestParams createHttpJsonRequestParams) {
    if (disposed) {
      throw new IllegalStateException("Object was disposed!");
    }

    HttpJsonRequest request = new HttpJsonRequest(createHttpJsonRequestParams, this);
    request.setShouldCacheRequest(createHttpJsonRequestParams.isAvoidCachingRequest() == false
        && createHttpJsonRequestParams.getConvention().getShouldCacheRequest().apply(createHttpJsonRequestParams.getUrl()));

    if (request.getShouldCacheRequest() && createHttpJsonRequestParams.getMethod() == HttpMethods.GET && !getDisableHttpCaching()) {
      //TODO: pass headers method
      CachedRequestOp cachedRequestDetails = configureCaching(createHttpJsonRequestParams.getUrl(), request);
      request.setCachedRequestDetails(cachedRequestDetails.getCachedRequest());
      request.setSkipServerCheck(cachedRequestDetails.isSkipServerCheck());
    }
    EventHelper.invoke(configureRequest, createHttpJsonRequestParams.getOwner(), new WebRequestEventArgs(request.getWebRequest()));
    return request;
  }

  public AutoCloseable disableAllCaching() {
    final Long oldAggressiveCaching = getAggressiveCacheDuration();
    final Boolean oldHttpCaching = getDisableHttpCaching();

    setAggressiveCacheDuration(null);
    setDisableHttpCaching(true);

    return new AutoCloseable() {
      @Override
      public void close() throws Exception {
        setAggressiveCacheDuration(oldAggressiveCaching);
        setDisableHttpCaching(oldHttpCaching);
      }
    };
  }

  public void expireItemsFromCache(String db)
  {
    cache.forceServerCheckOfCachedItemsForDatabase(db);
    numOfCacheResets++;
  }

  public Long getAggressiveCacheDuration() {
    return aggressiveCacheDuration.get();
  }

  RavenJToken getCachedResponse(HttpJsonRequest httpJsonRequest, Map<String, String> additionalHeaders) {
    if (httpJsonRequest.getCachedRequestDetails() == null) {
      throw new IllegalStateException("Cannot get cached response from a request that has no cached information");
    }
    httpJsonRequest.setResponseStatusCode(HttpStatus.SC_NOT_MODIFIED);
    httpJsonRequest.setResponseHeaders(new HashMap<String, String>(httpJsonRequest.getCachedRequestDetails().getHeaders()));

    if (additionalHeaders != null && additionalHeaders.containsKey(Constants.RAVEN_FORCE_PRIMARY_SERVER_CHECK)) {
      httpJsonRequest.getResponseHeaders().put(Constants.RAVEN_FORCE_PRIMARY_SERVER_CHECK, additionalHeaders.get(Constants.RAVEN_FORCE_PRIMARY_SERVER_CHECK));
    }

    incrementCachedRequests();
    return httpJsonRequest.getCachedRequestDetails().getData().cloneToken();
  }

  /**
   * The number of currently held requests in the cache
   * @return
   */
  public int getCurrentCacheSize() {
    return cache.getCurrentSize();
  }

  public boolean getDisableHttpCaching() {
    Boolean value = disableHttpCaching.get();
    if (value == null) {
      return false;
    }
    return disableHttpCaching.get();
  }


  /**
   * @return the numOfCachedRequests
   */
  public int getNumOfCachedRequests() {
    return numOfCachedRequests.get();
  }

  /**
   * @return the numOfCacheResets
   */
  public int getNumOfCacheResets() {
    return numOfCacheResets;
  }



  private void incrementCachedRequests() {
    numOfCachedRequests.incrementAndGet();
  }

  protected void invokeLogRequest(ServerClient sender, RequestResultArgs requestResult) {
    EventHelper.invoke(logRequest, sender, requestResult);
  }

  /**
   * @return the disableRequestCompression
   */
  public boolean isDisableRequestCompression() {
    return disableRequestCompression;
  }

  /**
   * @return the enableBasicAuthenticationOverUnsecuredHttpEvenThoughPasswordsWouldBeSentOverTheWireInClearTextToBeStolenByHackers
   */
  public boolean isEnableBasicAuthenticationOverUnsecuredHttpEvenThoughPasswordsWouldBeSentOverTheWireInClearTextToBeStolenByHackers() {
    return enableBasicAuthenticationOverUnsecuredHttpEvenThoughPasswordsWouldBeSentOverTheWireInClearTextToBeStolenByHackers;
  }

  public void removeConfigureRequestEventHandler(EventHandler<WebRequestEventArgs> event) {
    configureRequest.remove(event);
  }

  public void removeLogRequestEventHandler(EventHandler<RequestResultArgs> event) {
    logRequest.remove(event);
  }



  public void resetCache() {
    if (cache != null) {
      try {
        cache.close();
      } catch (Exception e) { /*ignore */ }
    }
    cache = new SimpleCache(maxNumberOfCachedRequests);
    numOfCachedRequests = new AtomicInteger();
  }

  public void setAggressiveCacheDuration(Long value) {
    aggressiveCacheDuration.set(value);
  }

  public void setDisableHttpCaching(Boolean value) {
    disableHttpCaching.set(value);
  }

  /**
   * @param disableRequestCompression the disableRequestCompression to set
   */
  public void setDisableRequestCompression(boolean disableRequestCompression) {
    this.disableRequestCompression = disableRequestCompression;
  }

  /**
   *  Advanced: Don't set this unless you know what you are doing!
   *  Enable using basic authentication using http
   *  By default, RavenDB only allows basic authentication over HTTPS, setting this property to true
   *  will instruct RavenDB to make unsecured calls (usually only good for testing / internal networks).
   * @param enableBasicAuthenticationOverUnsecuredHttpEvenThoughPasswordsWouldBeSentOverTheWireInClearTextToBeStolenByHackers the enableBasicAuthenticationOverUnsecuredHttpEvenThoughPasswordsWouldBeSentOverTheWireInClearTextToBeStolenByHackers to set
   */
  public void setEnableBasicAuthenticationOverUnsecuredHttpEvenThoughPasswordsWouldBeSentOverTheWireInClearTextToBeStolenByHackers(
      boolean enableBasicAuthenticationOverUnsecuredHttpEvenThoughPasswordsWouldBeSentOverTheWireInClearTextToBeStolenByHackers) {
    this.enableBasicAuthenticationOverUnsecuredHttpEvenThoughPasswordsWouldBeSentOverTheWireInClearTextToBeStolenByHackers = enableBasicAuthenticationOverUnsecuredHttpEvenThoughPasswordsWouldBeSentOverTheWireInClearTextToBeStolenByHackers;
  }

  /**
   * @param numOfCacheResets the numOfCacheResets to set
   */
  public void setNumOfCacheResets(int numOfCacheResets) {
    this.numOfCacheResets = numOfCacheResets;
  }

  protected void updateCacheTime(HttpJsonRequest httpJsonRequest) {
    if (httpJsonRequest.getCachedRequestDetails() == null) {
      throw new IllegalStateException("Cannot update cached response from a request that has no cached information");
    }
    httpJsonRequest.getCachedRequestDetails().setTime(new Date());
  }
}

File: Raven.Client.Lightweight/Connection/Implementation/HttpJsonRequestFactory.cs
using System.Net.Http;
using System;
using System.Collections.Specialized;
using System.Net;
using System.Net.Http.Headers;
using System.Threading;
using System.Runtime.Remoting.Messaging;

using Raven.Abstractions;
using Raven.Abstractions.Connection;
using Raven.Abstractions.Data;
using Raven.Abstractions.Extensions;
using Raven.Client.Connection.Implementation;
using Raven.Client.Connection.Profiling;
using Raven.Client.Extensions;
using Raven.Client.Util;
using Raven.Json.Linq;

namespace Raven.Client.Connection
{
    ///<summary>
    /// Create the HTTP Json Requests to the RavenDB Server
    /// and manages the http cache
    ///</summary>
    public class HttpJsonRequestFactory : IDisposable
    {
        /// <summary>
        /// Occurs when a json request is created
        /// </summary>
        public event EventHandler<WebRequestEventArgs> ConfigureRequest = delegate { };

        public event EventHandler OnDispose = delegate { }; 

        /// <summary>
        /// Occurs when a json request is completed
        /// </summary>
        public event EventHandler<RequestResultArgs> LogRequest;


        public bool CanLogRequest
        {
            get { return LogRequest != null; }
        }
        /// <summary>
        /// Invoke the LogRequest event
        /// </summary>
        internal void OnLogRequest(IHoldProfilingInformation sender, RequestResultArgs args)
        {
            var handler = LogRequest;
            if (handler != null)
                handler(sender, args);
        }

        private int maxNumberOfCachedRequests;

        internal readonly HttpClientCache httpClientCache;
        public HttpClientCache HttpClientCache
        {
            get { return httpClientCache; }
        }

        internal readonly Func<HttpMessageHandler> httpMessageHandler;

        internal readonly bool acceptGzipContent;

        internal readonly string authenticationScheme;

        private SimpleCache cache;

        internal int NumOfCachedRequests;

        /// <summary>
        /// Creates the HTTP json request.
        /// </summary>
        public HttpJsonRequest CreateHttpJsonRequest(CreateHttpJsonRequestParams createHttpJsonRequestParams)
        {
            if (disposed)
                throw new ObjectDisposedException(typeof(HttpJsonRequestFactory).FullName);

            if (RequestTimeout != null)
                createHttpJsonRequestParams.Timeout = RequestTimeout.Value;

            var request = new HttpJsonRequest(createHttpJsonRequestParams, this)
            {
                ShouldCacheRequest =
                    createHttpJsonRequestParams.AvoidCachingRequest == false &&
                    createHttpJsonRequestParams.Convention.ShouldCacheRequest(createHttpJsonRequestParams.Url)
            };

            if (request.ShouldCacheRequest && !DisableHttpCaching)
            {
                var cachedRequestDetails = ConfigureCaching(createHttpJsonRequestParams.Url, request.AddHeader);
                request.CachedRequestDetails = cachedRequestDetails.CachedRequest;
                request.SkipServerCheck = cachedRequestDetails.SkipServerCheck;
            }

            ConfigureRequest(createHttpJsonRequestParams.Owner, new WebRequestEventArgs { Client = request.httpClient, Credentials = createHttpJsonRequestParams.Credentials });
            return request;
        }

        internal CachedRequestOp ConfigureCaching(string url, Action<string, string> setHeader)
        {
            var cachedRequest = cache.Get(url);
            if (cachedRequest == null)
                return new CachedRequestOp { SkipServerCheck = false };
            bool skipServerCheck = false;
            if (AggressiveCacheDuration != null)
            {
                var duration = AggressiveCacheDuration.Value;
                if (duration.TotalSeconds > 0)
                    setHeader("Cache-Control", "max-age=" + duration.TotalSeconds);

                if (cachedRequest.ForceServerCheck == false && (SystemTime.UtcNow - cachedRequest.Time) < duration) // can serve directly from local cache
                    skipServerCheck = true;

                cachedRequest.ForceServerCheck = false;
            }

            setHeader("If-None-Match", cachedRequest.Headers[Constants.MetadataEtagField]);
            return new CachedRequestOp { SkipServerCheck = skipServerCheck, CachedRequest = cachedRequest };
        }


        /// <summary>
        /// Reset the number of cached requests and clear the entire cache
        /// Mostly used for testing
        /// </summary>
        public void ResetCache(int? newMaxNumberOfCachedRequests = null)
        {

            if (newMaxNumberOfCachedRequests != null && newMaxNumberOfCachedRequests.Value == maxNumberOfCachedRequests)
                return;

            if (cache != null)
                cache.Dispose();

            if (newMaxNumberOfCachedRequests != null)
            {
                maxNumberOfCachedRequests = newMaxNumberOfCachedRequests.Value;
            }
            cache = new SimpleCache(maxNumberOfCachedRequests);
            NumOfCachedRequests = 0;
        }

        public void ExpireItemsFromCache(string db)
        {
            cache.ForceServerCheckOfCachedItemsForDatabase(db);
            Interlocked.Increment(ref numberOfCacheResets);
        }

        /// <summary>
        /// The number of cache evictions forced by
        /// tracking changes if aggressive cache was enabled
        /// </summary>
        public int NumberOfCacheResets
        {
            get
            {
                return Thread.VolatileRead(ref numberOfCacheResets);
            }
        }

        /// <summary>
        /// The number of requests that we got 304 for 
        /// and were able to handle purely from the cache
        /// </summary>
        public int NumberOfCachedRequests
        {
            get { return NumOfCachedRequests; }
        }

        /// <summary>
        /// The number of currently held requests in the cache
        /// </summary>
        public int CurrentCacheSize
        {
            get { return cache.CurrentSize; }
        }

        /// <summary>
        /// Determine whether to use compression or not 
        /// </summary>
        public bool DisableRequestCompression { get; set; }

        /// <summary>
        /// default ctor
        /// </summary>
        /// <param name="maxNumberOfCachedRequests"></param>
        public HttpJsonRequestFactory(int maxNumberOfCachedRequests, Func<HttpMessageHandler> httpMessageHandler = null, bool acceptGzipContent = true, string authenticationScheme = null)
        {
            this.maxNumberOfCachedRequests = maxNumberOfCachedRequests;
            this.httpMessageHandler = httpMessageHandler;
            this.acceptGzipContent = acceptGzipContent;
            this.authenticationScheme = authenticationScheme;

            var maxIdleTime = ServicePointManager.MaxServicePointIdleTime;

            httpClientCache = new HttpClientCache(maxIdleTime);

            ResetCache();
        }

        ///<summary>
        /// The aggressive cache duration
        ///</summary>
        public TimeSpan? AggressiveCacheDuration
        {
            get
            {
                return CallContext.LogicalGetData("Raven/Client/AggressiveCacheDuration") as TimeSpan?;
            }
            set
            {
                CallContext.LogicalSetData("Raven/Client/AggressiveCacheDuration", value);
            }
        }

        ///<summary>
        /// Session timeout - Thread Local
        ///</summary>
        public TimeSpan? RequestTimeout
        {
            get
            {
                return CallContext.LogicalGetData("Raven/Client/RequestTimeout") as TimeSpan?;
            }
            set
            {
                CallContext.LogicalSetData("Raven/Client/RequestTimeout", value);
            }
        }

        /// <summary>
        /// Disable the HTTP caching
        /// </summary>
        public bool DisableHttpCaching
        {
            get
            {
                var value = CallContext.LogicalGetData("Raven/Client/DisableHttpCaching");
                if (value == null)
                    return false;

                return (bool)value;
            }
            set
            {
                CallContext.LogicalSetData("Raven/Client/DisableHttpCaching", value);
            }
        }

        /// <summary>
        /// Advanced: Don't set this unless you know what you are doing!
        /// 
        /// 
        /// Enable using basic authentication using http
        /// By default, RavenDB only allows basic authentication over HTTPS, setting this property to true
        /// will instruct RavenDB to make unsecured calls (usually only good for testing / internal networks).
        /// </summary>
        public bool EnableBasicAuthenticationOverUnsecuredHttpEvenThoughPasswordsWouldBeSentOverTheWireInClearTextToBeStolenByHackers { get; set; }

        private volatile bool disposed;
        private int numberOfCacheResets;

        internal RavenJToken GetCachedResponse(HttpJsonRequest httpJsonRequest, NameValueCollection additionalHeaders)
        {
            if (httpJsonRequest.CachedRequestDetails == null)
                throw new InvalidOperationException("Cannot get cached response from a request that has no cached information");
            httpJsonRequest.ResponseStatusCode = HttpStatusCode.NotModified;
            httpJsonRequest.ResponseHeaders = new NameValueCollection(httpJsonRequest.CachedRequestDetails.Headers);

            httpJsonRequest.ResponseHeaders.Remove(Constants.RavenForcePrimaryServerCheck);
            if (additionalHeaders != null && additionalHeaders[Constants.RavenForcePrimaryServerCheck] != null)
            {
                httpJsonRequest.ResponseHeaders.Set(Constants.RavenForcePrimaryServerCheck, additionalHeaders[Constants.RavenForcePrimaryServerCheck]);
            }

            IncrementCachedRequests();
            return httpJsonRequest.CachedRequestDetails.Data.CloneToken();
        }

        internal RavenJToken GetCachedResponse(HttpJsonRequest httpJsonRequest, HttpResponseHeaders additionalHeaders)
        {
            if (httpJsonRequest.CachedRequestDetails == null)
                throw new InvalidOperationException("Cannot get cached response from a request that has no cached information");
            httpJsonRequest.ResponseStatusCode = HttpStatusCode.NotModified;
            httpJsonRequest.ResponseHeaders = new NameValueCollection(httpJsonRequest.CachedRequestDetails.Headers);

            if (additionalHeaders != null)
            {
                string forcePrimaryServerCHeck = additionalHeaders.GetFirstValue(Constants.RavenForcePrimaryServerCheck);
                if (forcePrimaryServerCHeck != null)
                    httpJsonRequest.ResponseHeaders.Add(Constants.RavenForcePrimaryServerCheck, forcePrimaryServerCHeck);
            }

            IncrementCachedRequests();
            return httpJsonRequest.CachedRequestDetails.Data.CloneToken();
        }

        internal RavenJToken GetCachedResponse(HttpJsonRequest httpJsonRequest)
        {
            if (httpJsonRequest.CachedRequestDetails == null)
                throw new InvalidOperationException("Cannot get cached response from a request that has no cached information");
            httpJsonRequest.ResponseStatusCode = HttpStatusCode.NotModified;
            httpJsonRequest.ResponseHeaders = new NameValueCollection(httpJsonRequest.CachedRequestDetails.Headers);

            IncrementCachedRequests();
            return httpJsonRequest.CachedRequestDetails.Data.CloneToken();
        }

        internal void IncrementCachedRequests()
        {
            Interlocked.Increment(ref NumOfCachedRequests);
        }

        internal void CacheResponse(string url, RavenJToken data, NameValueCollection headers)
        {
            if (string.IsNullOrEmpty(headers[Constants.MetadataEtagField]))
                return;

            RavenJToken clone;
            if (data == null)
            {
                clone = null;
            }
            else
            {
                clone = data.CloneToken();
                clone.EnsureCannotBeChangeAndEnableSnapshotting();
            }

            cache.Set(url, new CachedRequest
            {
                Data = clone,
                Time = SystemTime.UtcNow,
                Headers = new NameValueCollection(headers),
                Database = MultiDatabase.GetDatabaseName(url)
            });
        }

        internal void CacheResponse(string url, RavenJToken data, HttpResponseHeaders headers)
        {
            if (headers.ETag == null || string.IsNullOrEmpty(headers.ETag.Tag))
                return;

            var clone = data.CloneToken();
            clone.EnsureCannotBeChangeAndEnableSnapshotting();

            cache.Set(url, new CachedRequest
            {
                Data = clone,
                Time = SystemTime.UtcNow,
                Headers = new NameValueCollection(),// TODO: Use headers
                Database = MultiDatabase.GetDatabaseName(url)
            });
        }

        /// <summary>
        /// Performs application-defined tasks associated with freeing, releasing, or resetting unmanaged resources.
        /// </summary>
        /// <filterpriority>2</filterpriority>
        public void Dispose()
        {
            if (disposed)
                return;
            disposed = true;
            cache.Dispose();
            httpClientCache.Dispose();
            OnDispose(this, EventArgs.Empty);
        }

        internal void UpdateCacheTime(HttpJsonRequest httpJsonRequest)
        {
            if (httpJsonRequest.CachedRequestDetails == null)
                throw new InvalidOperationException("Cannot update cached response from a request that has no cached information");
            httpJsonRequest.CachedRequestDetails.Time = SystemTime.UtcNow;
        }

        /// <summary>
        /// Disable all caching within the given scope
        /// </summary>
        public IDisposable DisableAllCaching()
        {
            var oldAggressiveCaching = AggressiveCacheDuration;
            var oldHttpCaching = DisableHttpCaching;

            AggressiveCacheDuration = null;
            DisableHttpCaching = true;

            return new DisposableAction(() =>
            {
                AggressiveCacheDuration = oldAggressiveCaching;
                DisableHttpCaching = oldHttpCaching;
            });
        }
    }
}
